libphal/phal_pdbg.C

```diff
@@ -323,4 +326,130 @@ bool hasControlTransitionedToHost()
+	if (locCode.length() < EXP_LOCATION_CODE_MIN_LENGTH) {
+		log(level::ERROR,
+		    "Given location code %s is not meet "
+		    "with minimum length %d",
+		    locCode.c_str(), EXP_LOCATION_CODE_MIN_LENGTH);
+		return std::nullopt;
+	}
+
+	// "-" should be there to seggregate all (FC, Node number and SE values)
+	// together.
+	// Note: Each (FC, Node number and SE) value can be seggregate by "."
+	// but, cec device tree just have unexpand format so, just skipping
+	// still first occurrence "-" and replacing with "fcs".
+	auto endPosOfFcs = locCode.find('-', EXP_LOCATION_CODE_MIN_LENGTH);
+	if (endPosOfFcs == std::string::npos) {
+		log(level::ERROR,
+		    "Given location code %s is not valid i.e could not find "
+		    "dash",
+		    locCode.c_str());
+		return std::nullopt;
+	}
+
+	std::string unExpandedLocCode("Ufcs");
+
+	if (locCode.length() > EXP_LOCATION_CODE_MIN_LENGTH) {
+		unExpandedLocCode.append(
+		    locCode.substr(endPosOfFcs, std::string::npos));
+	}
+
+	return unExpandedLocCode;
+}
+
+/**
+ * @brief Fetch FRU type from device tree.
+ *
+ * @details An api to get FRU ATTR_TYPE corresponding to
+ *          a given unexpanded location code.
+ *
+ * @param[in] locationCode - Location code in unexpanded format.
+ * @param[out] fruType - Fru type of a given location code.
+ *
+ * @return 0 if fru type is found.
+ * @return -1 if unable to find a fru with the given location code.
+ * @return DEVTREE_ATTR_READ_FAIL if unable to fetch ATTR_TYPE from device tree.
+ */
+static int fetchFruTypeFromDevTree(const LocationCode &locationCode,
+				   ATTR_TYPE_Type &fruType)
+{
+	ATTR_LOCATION_CODE_Type locCode;
+	// Defining a list of existing frus
+	// Note: We need to update the list if a new fru is added
+	std::vector<const char*> frus{"proc", "tpm", "dimm"};
+	struct pdbg_target *target;
+
+	for (auto fru : frus) {
+		pdbg_for_each_class_target(fru, target)
+		{
+			if (DT_GET_PROP(ATTR_LOCATION_CODE, target, locCode)) {
+				continue;
+			} else {
+				if (locCode == locationCode) {
+					if (DT_GET_PROP(ATTR_TYPE, target,
+							fruType)) {
+						return openpower::phal::
+						    exception::
+							DEVTREE_ATTR_READ_FAIL;
+					} else {
+						return 0; // success
+					}
+				}
+			}
+		}
+	}
+	return -1;
+}
+
+std::expected<FRUType, openpower::phal::exception::ERR_TYPE>
+    getFRUType(const LocationCode &locationCode) noexcept
+{
+	LocationCode unExpandedLocCode = locationCode;
+	if (!unExpandedLocCode.starts_with("Ufcs-")) {
+
+		auto loc = getUnexpandedLocCode(locationCode);
+		if (loc.has_value()) {
+			unExpandedLocCode = loc.value();
+		} else {
+			log(level::ERROR, "Given location code %s is not valid",
+			    unExpandedLocCode.c_str());
+			return std::unexpected(openpower::phal::exception::
+						   ATTR_LOC_CODE_NOT_VALID);
+		}
+	}
+	ATTR_TYPE_Type fruType = ENUM_ATTR_TYPE_NA;
+	auto ret = fetchFruTypeFromDevTree(unExpandedLocCode, fruType);
+	if (ret == 0) {
+		return fruType;
+	} else {
```

Comment:
Any reason to nest these if-else statements?  It would be better if we have if-elseif-else block instead.

Reply:
No particular reason, I thought i will keep success case in if and all the failure cases in else block.

Reply:
it is already evident from the code that they are dealing with failure scenarios. I would suggest to avoid nesting as it would reduce readability of the code and also makes it difficult to unit-test later.